autopush/bridge/apns_ping.py

```diff
@@ -0,0 +1,86 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from twisted.python import log
+import apns
+import time
+
+
+# https://github.com/djacobs/PyAPNs
+class APNSBridge(object):
+    apns = None
+    messages = {}
+    errors = {0: 'No error',
+              1: 'Processing error',
+              2: 'Missing device token',
+              3: 'Missing topic',
+              4: 'Missing payload',
+              5: 'Invalid token size',
+              6: 'Invalid topic size',
+              7: 'Invalid payload size',
+              8: 'Invalid token',
+              10: 'Shutdown',
+              255: 'Unknown',
+              }
+
+    def _connect(self):
+        self.apns = apns.APNs(use_sandbox=self.config.get("sandbox", False),
+                              cert_file=self.config.get("cert_file"),
+                              key_file=self.config.get("key_file"),
+                              enhanced=True)
+
+    def __init__(self, config, storage):
+        self.config = config
+        self.default_title = config.get("default_title", "SimplePush")
```

Comment:
not always. https://developer.apple.com/library/ios/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/Chapters/ApplePushService.html#//apple_ref/doc/uid/TP40008194-CH100-SW15 The title and body of the message are user displayable items and only appear if there's no content_available flag set. It may be a bug (because of the loc-* fields), but as of the latest 8.1 it works.